tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -788,76 +788,100 @@ class SPIRVEmitter : public ASTConsumer {
     }
   }
 
-  uint32_t doIntrinsicCallExpr(const CallExpr *callExpr) {
-    const FunctionDecl *callee = callExpr->getDirectCallee();
+  uint32_t processIntrinsicDot(const CallExpr *callExpr) {
     const uint32_t returnType =
         typeTranslator.translateType(callExpr->getType());
 
+    // Get the function parameters. Expect 2 vectors as parameters.
+    assert(callExpr->getNumArgs() == 2u);
+    const Expr *arg0 = callExpr->getArg(0);
+    const Expr *arg1 = callExpr->getArg(1);
+    const uint32_t arg0Id = doExpr(arg0);
+    const uint32_t arg1Id = doExpr(arg1);
+    QualType arg0Type = arg0->getType();
+    QualType arg1Type = arg1->getType();
+    const size_t vec0Size = hlsl::GetHLSLVecSize(arg0Type);
+    const size_t vec1Size = hlsl::GetHLSLVecSize(arg1Type);
+    const QualType vec0ComponentType = hlsl::GetHLSLVecElementType(arg0Type);
+    const QualType vec1ComponentType = hlsl::GetHLSLVecElementType(arg1Type);
+    assert(callExpr->getType() == vec1ComponentType);
+    assert(vec0ComponentType == vec1ComponentType);
+    assert(vec0Size == vec1Size);
+    assert(vec0Size >= 1 && vec0Size <= 4);
+
+    // According to HLSL reference, the dot function only works on integers
+    // and floats.
+    const auto returnTypeBuiltinKind =
+        cast<BuiltinType>(callExpr->getType().getTypePtr())->getKind();
+    assert(returnTypeBuiltinKind == BuiltinType::Float ||
+           returnTypeBuiltinKind == BuiltinType::Int ||
+           returnTypeBuiltinKind == BuiltinType::UInt);
+
+    // Special case: dot product of two vectors, each of size 1. That is
+    // basically the same as regular multiplication of 2 scalars.
+    if (vec0Size == 1) {
+      const spv::Op spvOp = translateOp(BO_Mul, arg0Type);
+      return theBuilder.createBinaryOp(spvOp, returnType, arg0Id, arg1Id);
+    }
+
+    // If the vectors are of type Float, we can use OpDot.
+    if (returnTypeBuiltinKind == BuiltinType::Float) {
+      return theBuilder.createBinaryOp(spv::Op::OpDot, returnType, arg0Id,
+                                       arg1Id);
+    }
+    // Vector component type is Integer (signed or unsigned).
+    // Create all instructions necessary to perform a dot product on
+    // two integer vectors. SPIR-V OpDot does not support integer vectors.
+    // Therefore, we use other SPIR-V instructions (addition and
+    // multiplication).
+    else {
+      uint32_t result = 0;
+      llvm::SmallVector<uint32_t, 4> multIds;
+
+      // Extract members from the two vectors and multiply them.
+      for (unsigned int i = 0; i < vec0Size; ++i) {
+        const uint32_t vec0member =
+            theBuilder.createCompositeExtract(returnType, arg0Id, {i});
+        const uint32_t vec1member =
+            theBuilder.createCompositeExtract(returnType, arg1Id, {i});
+        const spv::Op spvOp = translateOp(BO_Mul, arg0Type);
```

Comment:
Minor: This can be pulled out of the loop since it won't change. But this loop is <=4 iterations. So meh.